// compass-db/src/constants.js
export const COLLECTION_ALREADY_EXISTS = 'Collection already exists';
export const SOFT_DELETED_THE_COLLECTION = 'Soft-deleted the collection';
export const SOME_PROBLEM_OCCURED = 'Some problem occurred';
export const ERROR_IN_DROPING_COLLECTION = 'Error in collection';
export const ALREADY_INACTIVE_COLLECTION = 'Already inactive collection';
export const COLLECTION_DOESNT_EXIST = 'Collection doesn\'t exist';
export const LIST_UPDATED = 'list_of_all_collections.json updated';
export const ERROR_LIST_NOT_UPDATED = 'Error while operating list_of_all_collections.json file';
export const COLLECTION_NOT_ACTIVE = 'This Collection is not active';

export const DOCUMENT_ADDED = 'Document has been added to the collection';
export const ERROR_IN_ADDING_DOCUMENT = 'Some problem occured while adding the document';
export const DOCUMENT_FOUND = 'Found the document';
export const DOCUMENT_NOT_FOUND = 'Not able to find the document';
export const ERROR_IN_FINDING_DOCUMENT = 'Error in finding the document';
export const DOCUMENT_DELETED = 'Document has been deleted from the collection';
export const ERROR_IN_DELETING_DOCUMENT = 'Some problem occured while deleting the document';
export const DOCUMENT_UPDATED = 'Document has been updated from the collection';
export const ERROR_IN_UPDATING_DOCUMENT = 'Some problem occured while updating the document';

export const COLLECTION_MASTER = 'compassCollectionMaster.json';
