// blockstack/tests/unitTests/lib/sampleData.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.sampleAddressBasedVerifications = exports.sampleVerifications = exports.sampleProofs = exports.sampleTokenFiles = exports.sampleProfiles = exports.sampleNameRecords = exports.sampleManifests = undefined;

var _fs = require('fs');

var _fs2 = _interopRequireDefault(_fs);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var TEST_DATA_DIR = './tests/testData';

var sampleManifests = exports.sampleManifests = {
  helloBlockstack: {
    name: 'Hello, Blockstack',
    short_name: 'Hello, Blockstack',
    start_url: 'https://helloblockstack.com',
    display: 'standalone',
    background_color: '#fff',
    description: 'A simple app demonstrating how to log in with Blockstack.',
    icons: [{
      src: 'https://raw.githubusercontent.com/blockstack/blockstack-portal/master/app/images/app-hello-blockstack.png',
      sizes: '192x192',
      type: 'image/png'
    }]
  }
};

var sampleNameRecords = exports.sampleNameRecords = {
  ryan: JSON.parse(_fs2.default.readFileSync(TEST_DATA_DIR + '/name-records/ryan.json'))
};

var sampleProfiles = exports.sampleProfiles = {
  balloonDog: JSON.parse(_fs2.default.readFileSync(TEST_DATA_DIR + '/profiles/balloonDog.json')),
  naval: JSON.parse(_fs2.default.readFileSync(TEST_DATA_DIR + '/profiles/naval.json')),
  ryan: JSON.parse(_fs2.default.readFileSync(TEST_DATA_DIR + '/profiles/ryan.json')),
  larry: JSON.parse(_fs2.default.readFileSync(TEST_DATA_DIR + '/profiles/larry.json')),
  google: JSON.parse(_fs2.default.readFileSync(TEST_DATA_DIR + '/profiles/google.json')),
  navalLegacy: JSON.parse(_fs2.default.readFileSync(TEST_DATA_DIR + '/profiles/naval-legacy.json')),
  navalLegacyConvert: JSON.parse(_fs2.default.readFileSync(TEST_DATA_DIR + '/profiles/naval-legacy-convert.json'))
};

var sampleTokenFiles = exports.sampleTokenFiles = {
  ryan_apr20: {
    url: 'https://blockstack.s3.amazonaws.com/ryan_apr20.id',
    body: JSON.parse(_fs2.default.readFileSync(TEST_DATA_DIR + '/token-files/ryan_apr20.json'))
  },
  ryan: {
    url: 'https://blockstack.s3.amazonaws.com/ryan.id',
    body: JSON.parse(_fs2.default.readFileSync(TEST_DATA_DIR + '/token-files/ryan.json'), 'utf8')
  }
};

var sampleProofs = exports.sampleProofs = {
  naval: JSON.parse(_fs2.default.readFileSync(TEST_DATA_DIR + '/profiles/naval.proofs.json')),
  larry: JSON.parse(_fs2.default.readFileSync(TEST_DATA_DIR + '/profiles/larry.proofs.json')),
  ken: JSON.parse(_fs2.default.readFileSync(TEST_DATA_DIR + '/profiles/ken.proofs.json')),
  bruno: JSON.parse(_fs2.default.readFileSync(TEST_DATA_DIR + '/profiles/bruno.proofs.json'))
};

var sampleVerifications = exports.sampleVerifications = {
  naval: {
    facebook: {
      url: 'https://www.facebook.com/navalr/posts/10152190734077261',
      body: _fs2.default.readFileSync(TEST_DATA_DIR + '/profiles/naval.verification.facebook.html', 'utf8')
    },
    github: {
      url: 'https://gist.github.com/navalr/f31a74054f859ec0ac6a',
      body: _fs2.default.readFileSync(TEST_DATA_DIR + '/profiles/naval.verification.github.html', 'utf8')
    },
    twitter: {
      url: 'https://twitter.com/naval/status/486609266212499456',
      body: _fs2.default.readFileSync(TEST_DATA_DIR + '/profiles/naval.verification.twitter.html', 'utf8')
    }
  },
  larry: {
    facebook: {
      url: 'https://www.facebook.com/larry.salibra/posts/10100341028448093',
      body: _fs2.default.readFileSync(TEST_DATA_DIR + '/profiles/larry.verification.facebook.html', 'utf8')
    }
  }
};

var sampleAddressBasedVerifications = exports.sampleAddressBasedVerifications = {
  larry: {
    facebook: {
      url: 'https://www.facebook.com/larrysalibra/posts/10100341028448094',
      body: _fs2.default.readFileSync(TEST_DATA_DIR + '/profiles/larry.verification.address.facebook.html', 'utf8')
    }
  },
  ken: {
    github: {
      url: 'https://gist.github.com/yknl/37c763ab7bc6cf89b919212ef3f10676',
      body: _fs2.default.readFileSync(TEST_DATA_DIR + '/profiles/ken.verification.github.html', 'utf8')
    },
    twitter: {
      url: 'https://twitter.com/YukanL/status/903285763240022017',
      body: _fs2.default.readFileSync(TEST_DATA_DIR + '/profiles/ken.verification.twitter.html', 'utf8')
    },
    instagram: {
      url: 'https://www.instagram.com/p/BYj6UDwgaX7/',
      body: _fs2.default.readFileSync(TEST_DATA_DIR + '/profiles/ken.verification.instagram.html', 'utf8')
    },
    instagramRegression: {
      url: 'https://www.instagram.com/p/BYj6UDwgaX7/',
      body: _fs2.default.readFileSync(TEST_DATA_DIR + '/profiles/' + 'ken.verification.instagram.regression.html', 'utf8')
    },
    hackerNews: {
      url: 'https://news.ycombinator.com/user?id=yukanl',
      body: _fs2.default.readFileSync(TEST_DATA_DIR + '/profiles/ken.verification.hackernews.html', 'utf8')
    },
    linkedIn: {
      url: 'https://www.linkedin.com/feed/update/urn:li:activity:6311587377647222784/',
      body: _fs2.default.readFileSync(TEST_DATA_DIR + '/profiles/ken.verification.linkedin.html', 'utf8')
    },
    linkedInBroken: {
      url: 'https://www.linkedin.com/feed/update/urn:li:activity:6311587377647222784/',
      body: _fs2.default.readFileSync(TEST_DATA_DIR + '/profiles/ken.verification.linkedinbroken.html', 'utf8')
    }
  }
};
